File: kmp/shared/engine/src/commonTest/kotlin/game/vinto/engine/TossInStateTest.kt
package game.vinto.engine

import game.vinto.shapes.GameAction
import game.vinto.shapes.GameState
import game.vinto.shapes.GameSubPhase
import game.vinto.shapes.PlayerIdPayload
import game.vinto.shapes.Rank
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Whose turn it is during a toss-in, ported from
 * `packages/engine/src/lib/__tests__/toss-in-state.test.ts`.
 *
 * A toss-in window interrupts a turn without ending it, and that is the whole subject of this
 * file. `currentPlayerIndex` must keep pointing at whoever was playing until *everyone* has
 * said they are done — and if a queued action moves it to a tosser, it has to come back
 * afterwards. Get this wrong and the turn order quietly rotates, which shows up much later as
 * a player getting two turns in a row.
 */
class TossInStateTest {

    private fun finished(playerId: String) =
        GameAction.PlayerTossInFinished(PlayerIdPayload(playerId))

    private fun threeSeats(
        firstCards: List<game.vinto.shapes.Card>,
        secondCards: List<game.vinto.shapes.Card>,
        thirdCards: List<game.vinto.shapes.Card>,
    ) = listOf(
        testPlayer("human-1", "Human", isHuman = true, cards = firstCards),
        testPlayer("bot-1", "Bot 1", isHuman = false, cards = secondCards),
        testPlayer("bot-2", "Bot 2", isHuman = false, cards = thirdCards),
    )

    @Test
    fun theTurnStaysPutUntilEveryPlayerHasSaidTheyAreDone() {
        val state = testState(
            turnNumber = 1,
            subPhase = GameSubPhase.IDLE,
            players = threeSeats(
                listOf(testCard(Rank.SEVEN, "h1"), testCard(Rank.THREE, "h2")),
                listOf(testCard(Rank.SEVEN, "b1"), testCard(Rank.FOUR, "b2")),
                listOf(testCard(Rank.FIVE, "b3"), testCard(Rank.SIX, "b4")),
            ),
            drawPile = pileOf(
                testCard(Rank.TWO, "d1"), testCard(Rank.EIGHT, "d2"), testCard(Rank.NINE, "d3"),
            ),
        )

        var next = unsafeReduce(state, drawCard("human-1"))
        assertEquals(GameSubPhase.CHOOSING, next.subPhase)
        assertEquals(0, next.currentPlayerIndex)

        next = unsafeReduce(next, discardCard("human-1"))
        assertEquals(GameSubPhase.TOSS_QUEUE_ACTIVE, next.subPhase)
        assertNotNull(next.activeTossIn)
        assertEquals(0, next.currentPlayerIndex)

        next = unsafeReduce(next, participateInTossIn("bot-1", listOf(0)))
        assertEquals(0, next.currentPlayerIndex, "a toss-in must not move the turn")
        assertEquals(GameSubPhase.TOSS_QUEUE_ACTIVE, next.subPhase)

        next = unsafeReduce(next, finished("bot-1"))
        assertEquals(0, next.currentPlayerIndex)

        next = unsafeReduce(next, finished("bot-2"))
        assertEquals(0, next.currentPlayerIndex)

        // The last one to finish is what closes the window and moves the turn on.
        next = unsafeReduce(next, finished("human-1"))
        assertEquals(1, next.currentPlayerIndex)
        assertEquals(GameSubPhase.AI_THINKING, next.subPhase)
    }

    @Test
    fun botsCanTossIntoAWindowAHumanOpened() {
        val state = testState(
            turnNumber = 1,
            subPhase = GameSubPhase.TOSS_QUEUE_ACTIVE,
            players = threeSeats(
                listOf(testCard(Rank.TWO, "h1"), testCard(Rank.THREE, "h2")),
                listOf(testCard(Rank.KING, "b1k"), testCard(Rank.FOUR, "b2")),
                listOf(testCard(Rank.KING, "b2k"), testCard(Rank.SIX, "b4")),
            ),
            activeTossIn = tossIn(
                ranks = listOf(Rank.KING),
                initiatorId = "human-1",
                waitingForInput = true,
            ),
        )

        var next = unsafeReduce(state, participateInTossIn("bot-1", listOf(0)))
        assertEquals(1, next.activeTossIn?.queuedActions?.size)
        assertEquals("bot-1", next.activeTossIn?.queuedActions?.first()?.playerId)
        assertEquals(0, next.currentPlayerIndex)

        next = unsafeReduce(next, participateInTossIn("bot-2", listOf(0)))
        assertEquals(2, next.activeTossIn?.queuedActions?.size)
        assertEquals("bot-2", next.activeTossIn?.queuedActions?.get(1)?.playerId)
        assertEquals(0, next.currentPlayerIndex)

        next = markPlayersReady(next, listOf("bot-1", "bot-2", "human-1"))
        next = unsafeReduce(next, useCardAction("bot-1"))

        assertEquals(GameSubPhase.AWAITING_ACTION, next.subPhase)
        assertNotNull(next.pendingAction)
        assertEquals("bot-1", next.pendingAction?.playerId, "the queue is served in order")
    }

    @Test
    fun anAllBotWindowBehavesTheSameWay() {
        val state = testState(
            turnNumber = 1,
            subPhase = GameSubPhase.AI_THINKING,
            players = listOf(
                testPlayer(
                    "bot-1", "Bot 1", isHuman = false,
                    cards = listOf(testCard(Rank.QUEEN, "b1q"), testCard(Rank.THREE, "b1c2")),
                ),
                testPlayer(
                    "bot-2", "Bot 2", isHuman = false,
                    cards = listOf(testCard(Rank.TWO, "b2c1"), testCard(Rank.FOUR, "b2c2")),
                ),
                testPlayer(
                    "bot-3", "Bot 3", isHuman = false,
                    cards = listOf(testCard(Rank.FIVE, "b3c1"), testCard(Rank.SIX, "b3c2")),
                ),
            ),
            drawPile = pileOf(
                testCard(Rank.TWO, "d1"), testCard(Rank.EIGHT, "d2"), testCard(Rank.NINE, "d3"),
            ),
        )

        var next = unsafeReduce(state, drawCard("bot-1"))
        assertEquals(GameSubPhase.CHOOSING, next.subPhase)
        assertEquals(0, next.currentPlayerIndex)

        next = unsafeReduce(next, discardCard("bot-1"))
        assertEquals(GameSubPhase.TOSS_QUEUE_ACTIVE, next.subPhase)
        assertTrue(next.activeTossIn?.ranks?.contains(Rank.TWO) == true)
        assertEquals(0, next.currentPlayerIndex)

        next = unsafeReduce(next, participateInTossIn("bot-2", listOf(0)))
        assertEquals(0, next.activeTossIn?.queuedActions?.size, "a 2 has no action to queue")
        assertEquals(0, next.currentPlayerIndex)

        next = markPlayersReady(next, listOf("bot-1", "bot-2", "bot-3"))

        assertEquals(1, next.currentPlayerIndex)
        assertEquals(GameSubPhase.AI_THINKING, next.subPhase)
    }

    @Test
    fun aPlayerMayTossIntoTheirOwnWindowAndKeepTheirTurn() {
        val state = testState(
            turnNumber = 1,
            subPhase = GameSubPhase.TOSS_QUEUE_ACTIVE,
            players = threeSeats(
                listOf(testCard(Rank.SEVEN, "h1"), testCard(Rank.FOUR, "h2")),
                listOf(testCard(Rank.KING, "b1k"), testCard(Rank.FOUR, "b2")),
                listOf(testCard(Rank.KING, "b2k"), testCard(Rank.SIX, "b4")),
            ),
            activeTossIn = tossIn(
                ranks = listOf(Rank.SEVEN),
                initiatorId = "human-1",
                waitingForInput = true,
            ),
        )

        var next = unsafeReduce(state, participateInTossIn("human-1", listOf(0)))
        next = markPlayersReady(next, listOf("human-1", "bot-1", "bot-2"))

        assertEquals(1, next.activeTossIn?.queuedActions?.size)
        assertEquals("human-1", next.activeTossIn?.queuedActions?.first()?.playerId)
        assertEquals(0, next.currentPlayerIndex)

        next = unsafeReduce(next, selectTarget("human-1", "human-1", 0))
        next = unsafeReduce(next, confirmPeek("human-1"))

        assertEquals(GameSubPhase.TOSS_QUEUE_ACTIVE, next.subPhase)
        assertNull(next.pendingAction)
        assertEquals(0, next.currentPlayerIndex)
    }

    /**
     * The turn order has to come back.
     *
     * A queued toss-in action moves `currentPlayerIndex` to whoever tossed the card, so they
     * can aim it. Once the window closes it must return to the seat *after* the one whose
     * turn was interrupted — not to the tosser's neighbour. The three cases below take the
     * three different exits: a plain peek, a King declaring a card with no action, and a King
     * declaring one that has an action.
     */
    private fun humanSwapsIntoATossInThatBotTwoJoins(
        botTwoSecondCard: Rank,
        humanFirstCard: Rank,
        botTwoFirstCard: Rank,
    ): GameState {
        val state = testState(
            turnNumber = 1,
            subPhase = GameSubPhase.AI_THINKING,
            players = threeSeats(
                listOf(testCard(humanFirstCard, "h1"), testCard(Rank.FOUR, "h2")),
                listOf(testCard(Rank.KING, "b1k"), testCard(Rank.FOUR, "b2")),
                listOf(testCard(botTwoFirstCard, "b2k"), testCard(botTwoSecondCard, "b4")),
            ),
            drawPile = pileOf(testCard(Rank.TWO, "draw_2"), testCard(Rank.THREE, "draw_3")),
        )

        var next = unsafeReduce(state, drawCard("human-1"))
        next = unsafeReduce(next, swapCard("human-1", 0))
        next = unsafeReduce(next, participateInTossIn("bot-2", listOf(0)))

        assertEquals(1, next.players.first { it.id == "bot-2" }.cards.size)

        next = markPlayersReady(next, listOf("human-1", "bot-1", "bot-2"))

        assertEquals(1, next.activeTossIn?.queuedActions?.size)
        assertEquals("bot-2", next.activeTossIn?.queuedActions?.first()?.playerId)
        assertEquals(2, next.currentPlayerIndex, "the tosser takes the seat to aim their card")

        return unsafeReduce(next, selectTarget("bot-2", "bot-2", 0))
    }

    private fun assertTurnReturnedToBotOne(state: GameState, expectedTopRank: Rank) {
        assertEquals(1, state.currentPlayerIndex, "the turn did not return to the right seat")
        assertEquals(GameSubPhase.AI_THINKING, state.subPhase)
        assertNull(state.pendingAction)
        assertEquals(expectedTopRank, state.discardPile.peekTop()?.rank)
        assertEquals(false, state.discardPile.peekTop()?.played)
    }

    @Test
    fun aTossedInSevenReturnsTheTurnToTheInterruptedOrder() {
        var next = humanSwapsIntoATossInThatBotTwoJoins(
            botTwoSecondCard = Rank.SIX,
            humanFirstCard = Rank.SEVEN,
            botTwoFirstCard = Rank.SEVEN,
        )
        next = unsafeReduce(next, confirmPeek("bot-2"))
        next = markPlayersReady(next, listOf("human-1", "bot-1", "bot-2"))

        assertTurnReturnedToBotOne(next, Rank.SEVEN)
    }

    @Test
    fun aTossedInKingDeclaringANonActionCardReturnsTheTurnToo() {
        var next = humanSwapsIntoATossInThatBotTwoJoins(
            botTwoSecondCard = Rank.SIX,
            humanFirstCard = Rank.KING,
            botTwoFirstCard = Rank.KING,
        )
        next = unsafeReduce(next, declareKing("bot-2", Rank.SIX))
        next = markPlayersReady(next, listOf("human-1", "bot-1", "bot-2"))

        assertTurnReturnedToBotOne(next, Rank.KING)
    }

    @Test
    fun aTossedInKingDeclaringAnActionCardReturnsTheTurnToo() {
        var next = humanSwapsIntoATossInThatBotTwoJoins(
            botTwoSecondCard = Rank.SEVEN,
            humanFirstCard = Rank.KING,
            botTwoFirstCard = Rank.KING,
        )
        next = unsafeReduce(next, declareKing("bot-2", Rank.SEVEN))
        next = markPlayersReady(next, listOf("human-1", "bot-1", "bot-2"))

        assertTurnReturnedToBotOne(next, Rank.KING)
    }
}

File: legacy-web/packages/engine/src/lib/__tests__/toss-in-state.test.ts
// Test to verify toss-in state behavior, especially currentPlayerIndex
// This test verifies that during toss-in phase, currentPlayerIndex remains
// at the player who triggered the toss-in, not advancing until toss-in completes

import { describe, it, expect } from 'vitest';
import { GameEngine } from '../game-engine';
import { GameActions } from '../game-actions';
import {
  createTestCard,
  createTestPlayer,
  createTestState,
  markPlayersReady,
  toPile,
  unsafeReduce,
} from './test-helpers';
import { Pile } from '@vinto/shapes';

describe('GameEngine - Toss-In State Management', () => {
  it('should maintain currentPlayerIndex during toss-in phase until all players ready', () => {
    // Setup: Player 0 (human) has a 7, Player 1 (bot) has a 7
    const state = createTestState({
      currentPlayerIndex: 0,
      turnNumber: 1,
      subPhase: 'idle',
      players: [
        createTestPlayer('human-1', 'Human', true, [
          createTestCard('7', 'h1'),
          createTestCard('3', 'h2'),
        ]),
        createTestPlayer('bot-1', 'Bot 1', false, [
          createTestCard('7', 'b1'),
          createTestCard('4', 'b2'),
        ]),
        createTestPlayer('bot-2', 'Bot 2', false, [
          createTestCard('5', 'b3'),
          createTestCard('6', 'b4'),
        ]),
      ],
      drawPile: toPile([
        createTestCard('2', 'd1'),
        createTestCard('8', 'd2'),
        createTestCard('9', 'd3'),
      ]),
    });

    // Human draws a card
    let newState = unsafeReduce(state, GameActions.drawCard('human-1'));
    expect(newState.subPhase).toBe('choosing');
    expect(newState.currentPlayerIndex).toBe(0); // Still human's turn

    // Human discards the drawn card (triggers toss-in)
    newState = unsafeReduce(newState, GameActions.discardCard('human-1'));

    // Verify toss-in state
    expect(newState.subPhase).toBe('toss_queue_active');
    expect(newState.activeTossIn).not.toBeNull();
    expect(newState.currentPlayerIndex).toBe(0); // CRITICAL: Still pointing to human!

    // Bot 1 tosses in their 7
    newState = unsafeReduce(
      newState,
      GameActions.participateInTossIn('bot-1', [0])
    );
    expect(newState.currentPlayerIndex).toBe(0); // STILL pointing to human
    expect(newState.subPhase).toBe('toss_queue_active');

    // Bot 1 marks ready
    newState = unsafeReduce(
      newState,
      GameActions.playerTossInFinished('bot-1')
    );
    expect(newState.currentPlayerIndex).toBe(0); // STILL pointing to human

    // Bot 2 marks ready
    newState = unsafeReduce(
      newState,
      GameActions.playerTossInFinished('bot-2')
    );
    expect(newState.currentPlayerIndex).toBe(0); // STILL pointing to human

    // Human marks ready - this should trigger turn advancement
    newState = unsafeReduce(
      newState,
      GameActions.playerTossInFinished('human-1')
    );

    // Now currentPlayerIndex should advance to next player (bot-1)
    expect(newState.currentPlayerIndex).toBe(1);
    expect(newState.subPhase).toBe('ai_thinking'); // Bot's turn
  });

  it('should allow bots to participate in toss-in when human triggers it', () => {
    // Setup: Player 0 (human) discards a King, Players 1-2 (bots) have Kings
    const state = createTestState({
      currentPlayerIndex: 0,
      turnNumber: 1,
      subPhase: 'toss_queue_active',
      players: [
        createTestPlayer('human-1', 'Human', true, [
          createTestCard('2', 'h1'),
          createTestCard('3', 'h2'),
        ]),
        createTestPlayer('bot-1', 'Bot 1', false, [
          createTestCard('K', 'b1k'),
          createTestCard('4', 'b2'),
        ]),
        createTestPlayer('bot-2', 'Bot 2', false, [
          createTestCard('K', 'b2k'),
          createTestCard('6', 'b4'),
        ]),
      ],
      activeTossIn: {
        ranks: ['K'],
        initiatorId: 'human-1',
        originalPlayerIndex: 0,
        participants: [],
        queuedActions: [],
        waitingForInput: true,
        playersReadyForNextTurn: [],
      },
    });

    // Verify initial state
    expect(state.currentPlayerIndex).toBe(0); // Human's turn
    expect(state.subPhase).toBe('toss_queue_active');

    // Bot 1 should be able to toss in
    let newState = unsafeReduce(
      state,
      GameActions.participateInTossIn('bot-1', [0])
    );
    expect(newState.activeTossIn?.queuedActions.length).toBe(1);
    expect(newState.activeTossIn?.queuedActions[0].playerId).toBe('bot-1');
    expect(newState.currentPlayerIndex).toBe(0); // Still human's turn

    // Bot 2 should be able to toss in
    newState = unsafeReduce(
      newState,
      GameActions.participateInTossIn('bot-2', [0])
    );
    expect(newState.activeTossIn?.queuedActions.length).toBe(2);
    expect(newState.activeTossIn?.queuedActions[1].playerId).toBe('bot-2');
    expect(newState.currentPlayerIndex).toBe(0); // STILL human's turn

    // All players mark ready
    newState = unsafeReduce(
      newState,
      GameActions.playerTossInFinished('bot-1')
    );
    newState = unsafeReduce(
      newState,
      GameActions.playerTossInFinished('bot-2')
    );
    newState = unsafeReduce(
      newState,
      GameActions.playerTossInFinished('human-1')
    );

    newState = unsafeReduce(newState, GameActions.playCardAction('bot-1'));

    // Should now be processing queued actions
    expect(newState.subPhase).toBe('awaiting_action');
    expect(newState.pendingAction).not.toBeNull();
    expect(newState.pendingAction?.playerId).toBe('bot-1'); // First queued action
  });

  it('should handle all-bot toss-in scenario correctly', () => {
    // This is the critical test: What happens when a bot triggers toss-in
    // and other bots need to participate?
    const state = createTestState({
      currentPlayerIndex: 0,
      turnNumber: 1,
      subPhase: 'ai_thinking', // Bot's turn
      players: [
        createTestPlayer('bot-1', 'Bot 1', false, [
          createTestCard('Q', 'b1q'),
          createTestCard('3', 'b1c2'),
        ]),
        createTestPlayer('bot-2', 'Bot 2', false, [
          createTestCard('2', 'b2c1'), // Has matching '2'
          createTestCard('4', 'b2c2'),
        ]),
        createTestPlayer('bot-3', 'Bot 3', false, [
          createTestCard('5', 'b3c1'),
          createTestCard('6', 'b3c2'),
        ]),
      ],
      drawPile: toPile([
        createTestCard('2', 'd1'), // This will be drawn
        createTestCard('8', 'd2'),
        createTestCard('9', 'd3'),
      ]),
    });

    // Bot 1 draws a '2'
    let newState = unsafeReduce(state, GameActions.drawCard('bot-1'));
    expect(newState.subPhase).toBe('choosing');
    expect(newState.currentPlayerIndex).toBe(0);

    // Bot 1 discards the '2' (triggers toss-in for rank '2')
    newState = unsafeReduce(newState, GameActions.discardCard('bot-1'));

    // Verify toss-in state
    expect(newState.subPhase).toBe('toss_queue_active');
    expect(newState.activeTossIn).not.toBeNull();
    expect(newState.activeTossIn?.ranks).toContain('2');
    expect(newState.currentPlayerIndex).toBe(0); // Still bot-1

    // Bot 2 tosses in their '2'
    newState = unsafeReduce(
      newState,
      GameActions.participateInTossIn('bot-2', [0])
    );
    expect(newState.activeTossIn?.queuedActions.length).toBe(0); // '2' is not an action card
    expect(newState.currentPlayerIndex).toBe(0); // STILL bot-1

    // All bots mark ready
    newState = unsafeReduce(
      newState,
      GameActions.playerTossInFinished('bot-1')
    );
    newState = unsafeReduce(
      newState,
      GameActions.playerTossInFinished('bot-2')
    );
    newState = unsafeReduce(
      newState,
      GameActions.playerTossInFinished('bot-3')
    );

    // Turn should advance to bot-2, no queued actions
    expect(newState.currentPlayerIndex).toBe(1);
    expect(newState.subPhase).toBe('ai_thinking'); // Bot-2's turn, no actions queued
  });

  it('should allow to say vinto after toss in', () => {
    // Setup: Player 0 (human) discards a King, Players 1-2 (bots) have Kings
    const state = createTestState({
      currentPlayerIndex: 0,
      turnNumber: 1,
      subPhase: 'toss_queue_active',
      players: [
        createTestPlayer('human-1', 'Human', true, [
          createTestCard('7', 'h1'),
          createTestCard('4', 'h2'),
        ]),
        createTestPlayer('bot-1', 'Bot 1', false, [
          createTestCard('K', 'b1k'),
          createTestCard('4', 'b2'),
        ]),
        createTestPlayer('bot-2', 'Bot 2', false, [
          createTestCard('K', 'b2k'),
          createTestCard('6', 'b4'),
        ]),
      ],
      activeTossIn: {
        ranks: ['7'],
        initiatorId: 'human-1',
        originalPlayerIndex: 0,
        participants: [],
        queuedActions: [],
        waitingForInput: true,
        playersReadyForNextTurn: [],
      },
    });

    // Verify initial state
    expect(state.currentPlayerIndex).toBe(0); // Human's turn
    expect(state.subPhase).toBe('toss_queue_active');

    let newState = unsafeReduce(
      state,
      GameActions.participateInTossIn('human-1', [0])
    );

    newState = markPlayersReady(newState, ['human-1', 'bot-1', 'bot-2']);

    expect(newState.activeTossIn?.queuedActions.length).toBe(1);
    expect(newState.activeTossIn?.queuedActions[0].playerId).toBe('human-1');
    expect(newState.currentPlayerIndex).toBe(0); // STILL human's turn

    // human plays his toss in card
    newState = unsafeReduce(
      newState,
      GameActions.selectActionTarget('human-1', 'human-1', 0)
    );
    newState = unsafeReduce(newState, GameActions.confirmPeek('human-1'));

    // Should now be processing queued actions
    expect(newState.subPhase).toBe('toss_queue_active');
    expect(newState.pendingAction).toBeNull();
    expect(newState.currentPlayerIndex).toBe(0); // STILL human's turn
  });

  it('should allow play non-King card drawn before toss in', () => {
    // Setup: Player 0 (human) discards Seven, Player 2 (bot) toss in Seven and play it
    // Player 1 should be able to play the card drawn before toss in
    const state = createTestState({
      currentPlayerIndex: 0,
      turnNumber: 1,
      subPhase: 'ai_thinking',
      players: [
        createTestPlayer('human-1', 'Human', true, [
          createTestCard('7', 'h1'),
          createTestCard('4', 'h2'),
        ]),
        createTestPlayer('bot-1', 'Bot 1', false, [
          createTestCard('K', 'b1k'),
          createTestCard('4', 'b2'),
        ]),
        createTestPlayer('bot-2', 'Bot 2', false, [
          createTestCard('7', 'b2k'),
          createTestCard('6', 'b4'),
        ]),
      ],
      drawPile: Pile.fromCards([
        createTestCard('2', 'draw_2'),
        createTestCard('3', 'draw_3'),
      ]),
    });

    // Verify initial state
    expect(state.currentPlayerIndex).toBe(0); // Human's turn

    let newState = unsafeReduce(state, GameActions.drawCard('human-1'));

    newState = unsafeReduce(newState, GameActions.swapCard('human-1', 0));

    newState = unsafeReduce(
      newState,
      GameActions.participateInTossIn('bot-2', [0])
    );

    newState = markPlayersReady(newState, ['human-1', 'bot-1', 'bot-2']);

    // Should now be processing queued actions
    expect(newState.activeTossIn?.queuedActions.length).toBe(1);
    expect(newState.activeTossIn?.queuedActions[0].playerId).toBe('bot-2');
    expect(newState.currentPlayerIndex).toBe(2);

    // bot-2 plays his toss in card
    newState = unsafeReduce(
      newState,
      GameActions.selectActionTarget('bot-2', 'bot-2', 0)
    );
    newState = unsafeReduce(newState, GameActions.confirmPeek('bot-2'));

    newState = markPlayersReady(newState, ['human-1', 'bot-1', 'bot-2']);

    // restoring original player order after toss in finished
    expect(newState.currentPlayerIndex).toBe(1);
    expect(newState.subPhase).toBe('ai_thinking'); // bot-1's turn
    expect(newState.pendingAction).toBeNull();
    expect(newState.currentPlayerIndex).toBe(1);
    expect(newState.discardPile.peekTop()?.rank).toBe('7'); // bot-2's played card
    expect(newState.discardPile.peekTop()?.played).toBe(false); // bot-1 has drawn card before toss in without playing
  });

  it('should allow play King card with declaring non-actionable card with non drawn before toss in', () => {
    // Setup: Player 0 (human) discards King, Player 2 (bot) toss in King and play it, declaring non-actionable card
    // Player 1 should be able to play the card drawn before toss in
    const state = createTestState({
      currentPlayerIndex: 0,
      turnNumber: 1,
      subPhase: 'ai_thinking',
      players: [
        createTestPlayer('human-1', 'Human', true, [
          createTestCard('K', 'h1'),
          createTestCard('4', 'h2'),
        ]),
        createTestPlayer('bot-1', 'Bot 1', false, [
          createTestCard('K', 'b1k'),
          createTestCard('4', 'b2'),
        ]),
        createTestPlayer('bot-2', 'Bot 2', false, [
          createTestCard('K', 'b2k'),
          createTestCard('6', 'b4'),
        ]),
      ],
      drawPile: Pile.fromCards([
        createTestCard('2', 'draw_2'),
        createTestCard('3', 'draw_3'),
      ]),
    });

    // Verify initial state
    expect(state.currentPlayerIndex).toBe(0); // Human's turn

    let newState = unsafeReduce(state, GameActions.drawCard('human-1'));

    newState = unsafeReduce(newState, GameActions.swapCard('human-1', 0));

    newState = unsafeReduce(
      newState,
      GameActions.participateInTossIn('bot-2', [0])
    );

    expect(newState.players.find((p) => p.id === 'bot-2')?.cards.length).toBe(
      1
    ); // only non-King left in hand
    newState = markPlayersReady(newState, ['human-1', 'bot-1', 'bot-2']);

    // Should now be processing queued actions
    expect(newState.activeTossIn?.queuedActions.length).toBe(1);
    expect(newState.activeTossIn?.queuedActions[0].playerId).toBe('bot-2');
    expect(newState.currentPlayerIndex).toBe(2);

    // bot-2 plays his toss in card
    newState = unsafeReduce(
      newState,
      GameActions.selectActionTarget('bot-2', 'bot-2', 0)
    );
    newState = unsafeReduce(
      newState,
      GameActions.declareKingAction('bot-2', '6')
    );

    newState = markPlayersReady(newState, ['human-1', 'bot-1', 'bot-2']);

    // restoring original player order after toss in finished
    expect(newState.currentPlayerIndex).toBe(1);
    expect(newState.subPhase).toBe('ai_thinking'); // bot-1's turn
    expect(newState.pendingAction).toBeNull();
    expect(newState.currentPlayerIndex).toBe(1);
    expect(newState.discardPile.peekTop()?.rank).toBe('K'); // human's unplayed King card
    expect(newState.discardPile.peekTop()?.played).toBe(false); // bot-1 has drawn card before toss in without playing
  });

  it('should allow play King card with declaring actionable card with non drawn before toss in', () => {
    // Setup: Player 0 (human) discards King, Player 2 (bot) toss in King and play it, declaring actionable card
    // Player 1 should be able to play the card drawn before toss in
    const state = createTestState({
      currentPlayerIndex: 0,
      turnNumber: 1,
      subPhase: 'ai_thinking',
      players: [
        createTestPlayer('human-1', 'Human', true, [
          createTestCard('K', 'h1'),
          createTestCard('4', 'h2'),
        ]),
        createTestPlayer('bot-1', 'Bot 1', false, [
          createTestCard('K', 'b1k'),
          createTestCard('4', 'b2'),
        ]),
        createTestPlayer('bot-2', 'Bot 2', false, [
          createTestCard('K', 'b2k'),
          createTestCard('7', 'b4'),
        ]),
      ],
      drawPile: Pile.fromCards([
        createTestCard('2', 'draw_2'),
        createTestCard('3', 'draw_3'),
      ]),
    });

    // Verify initial state
    expect(state.currentPlayerIndex).toBe(0); // Human's turn

    let newState = unsafeReduce(state, GameActions.drawCard('human-1'));

    newState = unsafeReduce(newState, GameActions.swapCard('human-1', 0));

    newState = unsafeReduce(
      newState,
      GameActions.participateInTossIn('bot-2', [0])
    );

    expect(newState.players.find((p) => p.id === 'bot-2')?.cards.length).toBe(
      1
    ); // only non-King left in hand
    newState = markPlayersReady(newState, ['human-1', 'bot-1', 'bot-2']);

    // Should now be processing queued actions
    expect(newState.activeTossIn?.queuedActions.length).toBe(1);
    expect(newState.activeTossIn?.queuedActions[0].playerId).toBe('bot-2');
    expect(newState.currentPlayerIndex).toBe(2);

    // bot-2 plays his toss in card
    newState = unsafeReduce(
      newState,
      GameActions.selectActionTarget('bot-2', 'bot-2', 0)
    );
    newState = unsafeReduce(
      newState,
      GameActions.declareKingAction('bot-2', '7')
    );

    newState = markPlayersReady(newState, ['human-1', 'bot-1', 'bot-2']);

    // restoring original player order after toss in finished
    expect(newState.currentPlayerIndex).toBe(1);
    expect(newState.subPhase).toBe('ai_thinking'); // bot-1's turn
    expect(newState.pendingAction).toBeNull();
    expect(newState.currentPlayerIndex).toBe(1);
    expect(newState.discardPile.peekTop()?.rank).toBe('K'); // human's unplayed King card
    expect(newState.discardPile.peekTop()?.played).toBe(false); // bot-1 has drawn card before toss in without playing
  });
});
